packages/flutter/lib/src/rendering/wrap.dart

```diff
@@ -559,40 +648,96 @@ class RenderWrap extends RenderBox
       size = constraints.smallest;
       return;
     }
-    final BoxConstraints childConstraints;
+    final BoxConstraints childConstraintsLoose;
+    final BoxConstraints childConstraintsTight;
     double mainAxisLimit = 0.0;
     bool flipMainAxis = false;
     bool flipCrossAxis = false;
+    final double spacing = this.spacing;
+    final double runSpacing = this.runSpacing;
+    final double Function(RenderBox child, double crossAxisSize) getChildMinIntrinsicMainAxisExtent;
+    final BoxConstraints Function(double maxSize) childConstraintsFittingLooseInRun;
+    final BoxConstraints Function(double maxSize) childConstraintsFittingTightInRun;
     switch (direction) {
       case Axis.horizontal:
-        childConstraints = BoxConstraints(maxWidth: constraints.maxWidth);
+        childConstraintsLoose = BoxConstraints(maxWidth: constraints.maxWidth);
+        childConstraintsTight = BoxConstraints(minWidth: constraints.maxWidth, maxWidth: constraints.maxWidth);
         mainAxisLimit = constraints.maxWidth;
+        getChildMinIntrinsicMainAxisExtent = (RenderBox child, double crossAxisSize) => child.getMinIntrinsicWidth(crossAxisSize);
+        childConstraintsFittingLooseInRun = (double maxSize) => BoxConstraints(maxWidth: maxSize);
+        childConstraintsFittingTightInRun = (double maxSize) => BoxConstraints(minWidth: maxSize, maxWidth: maxSize);
         if (textDirection == TextDirection.rtl) {
           flipMainAxis = true;
         }
         if (verticalDirection == VerticalDirection.up) {
           flipCrossAxis = true;
         }
       case Axis.vertical:
-        childConstraints = BoxConstraints(maxHeight: constraints.maxHeight);
+        childConstraintsLoose = BoxConstraints(maxHeight: constraints.maxHeight);
+        childConstraintsTight = BoxConstraints(minHeight: constraints.maxHeight, maxHeight: constraints.maxHeight);
         mainAxisLimit = constraints.maxHeight;
+        getChildMinIntrinsicMainAxisExtent = (RenderBox child, double crossAxisSize) => child.getMinIntrinsicHeight(crossAxisSize);
+        childConstraintsFittingLooseInRun = (double maxSize) => BoxConstraints(maxHeight: maxSize);
+        childConstraintsFittingTightInRun = (double maxSize) => BoxConstraints(minHeight: maxSize, maxHeight: maxSize);
         if (verticalDirection == VerticalDirection.up) {
           flipMainAxis = true;
         }
         if (textDirection == TextDirection.rtl) {
           flipCrossAxis = true;
         }
     }
-    final double spacing = this.spacing;
-    final double runSpacing = this.runSpacing;
     final List<_RunMetrics> runMetrics = <_RunMetrics>[];
     double mainAxisExtent = 0.0;
     double crossAxisExtent = 0.0;
     double runMainAxisExtent = 0.0;
     double runCrossAxisExtent = 0.0;
     int childCount = 0;
     while (child != null) {
-      child.layout(childConstraints, parentUsesSize: true);
+      final WrapParentData childParentData = child.parentData! as WrapParentData;
```

Comment:
Many of the same comments I made above also apply to this code.

Reply:
(I wonder if we could find a way here to share the code a little better)

Reply:
I'm pretty sure there was a common logic for `computeDryLayout` and `performLayout` in the past. I've introduced it in 9b67d01660.